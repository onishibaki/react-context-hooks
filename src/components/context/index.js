import { createContext, useContext } from "react";

export const EmployeesContext = createContext();

export const useEmployeeState = () => useContext(EmployeesContext);
