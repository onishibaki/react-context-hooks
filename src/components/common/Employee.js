import React, { useState } from "react";
import { useEmployeeState } from "../context";
import styled from "styled-components";

import { ActionTypes, addEmployee, updateEmployee } from "../action";

const EmployeeContainer = styled.div`
  border: 1px solid black;
  background: rgba(0, 0, 0, 0.5);
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
  text-align: center;
`;

const EmployeeWrapper = styled.div`
  border: 1px solid rgb(17, 17, 17);
  background-color: white;
  margin-top: 158px;
  display: inline-block;
  padding: 35px 10px 10px 10px;
  position: relative;
`;
const Button = styled.button`
  text-decoration: none;
  border: 1px solid #000000;
  padding: 10px 20px;
  font-weight: 700;
  color: black;
  font-size: 12px;
  margin: 10px 0;
  display: block;
  &:hover {
    background: #00000059;
    color: white;
  }
`;
const SpanButton = styled.span`
  font-weight: 700;
  color: black;
  font-size: 15px;
  margin: 10px 0;
  display: inline-block;
  position: absolute;
  top: -8px;
  right: 9px;
  cursor: pointer;
}
`;

const Input = styled.input`
  margin: 0 10px;
}
`;

export const Employee = () => {
  //Use Context
  const [
    {
      form,
      employee: { firstName, lastName, employee_id: employeeId },
      formMessage,
    },
    dispatch,
  ] = useEmployeeState();

  //Set state for First Name and Last Name Changes
  const [stateFirstName, setStateFirstName] = useState("");
  const [stateLastName, setStateLastName] = useState("");

  //Set state for setting blank value
  const [stateUpdateFirstName, setStateUpdateFirstName] = useState(null);
  const [stateUpdateLastNamet, setStateUpdateLastName] = useState(null);

  //Set the First Value
  let firstNameParameter = firstName;
  let lastNameParameter = lastName;

  //Parameter for Save
  const employeeParamSave = {
    stateFirstName,
    stateLastName,
  };

  //Checking for State if First name and Last name is blank to throw error
  if (stateUpdateFirstName === "") {
    firstNameParameter = stateUpdateFirstName;
  }
  if (stateUpdateLastNamet === "") {
    lastNameParameter = stateUpdateLastNamet;
  }

  //Parameter for Update
  const employeeParamUpdate = {
    employeeId,
    stateFirstName: stateFirstName || firstNameParameter,
    stateLastName: stateLastName || lastNameParameter,
  };

  //Getting the value of Input changes
  const inputChangedHandlerFirstName = (event) => {
    setStateFirstName(event.target.value);
    setStateUpdateFirstName("");
  };

  const inputChangedHandlerLastName = (event) => {
    setStateLastName(event.target.value);
    setStateUpdateLastName("");
  };

  return (
    //Condition to display Save and Update Screen
    form !== "" &&
    form !== "view" && (
      <EmployeeContainer className="employee-container">
        <EmployeeWrapper>
          <SpanButton
            onClick={() => {
              dispatch({
                type: ActionTypes.SET_FORM_STATE,
                form: "",
                data: {},
              });
              dispatch({
                type: ActionTypes.FORM_MESSAGE,
                data: "",
              });
              setStateLastName("");
              setStateFirstName("");
              setStateUpdateFirstName(null);
              setStateUpdateLastName(null);
            }}
          >
            X
          </SpanButton>
          <label>First Name: </label>
          <Input
            name="fname"
            defaultValue={firstName}
            onChange={inputChangedHandlerFirstName}
          ></Input>
          <label>Last Name: </label>
          <Input
            name="lname"
            defaultValue={lastName}
            onChange={inputChangedHandlerLastName}
          ></Input>
          {Object.keys(formMessage).length !== 0 && (
            <h2>{formMessage.message}</h2>
          )}
          {form !== "update" ? (
            <Button
              onClick={() => {
                addEmployee(employeeParamSave).then(function (response) {
                  dispatch(response);
                });
              }}
            >
              Save
            </Button>
          ) : (
            <Button
              onClick={() => {
                updateEmployee(employeeParamUpdate).then(function (response) {
                  dispatch(response);
                });
              }}
            >
              Update
            </Button>
          )}
        </EmployeeWrapper>
      </EmployeeContainer>
    )
  );
};
