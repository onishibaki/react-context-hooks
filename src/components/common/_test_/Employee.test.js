import React from "react";
import { render, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import Adapter from "enzyme-adapter-react-16";
import { configure, mount } from "enzyme";

import * as EmployeesContext from "../../context";
import { Employee } from "../Employee";

configure({ adapter: new Adapter() });

describe("Employee Component Test", () => {
  let setValue;
  let container;
  beforeEach(() => {
    setValue = (dataParameter) => {
      jest
        .spyOn(EmployeesContext, "useEmployeeState")
        .mockImplementation(() => dataParameter);
    };
    container = document.createElement("div");
  });

  afterEach(cleanup);

  it("Feedback Component Renders Correctly", () => {
    const data = [
      {
        form: "save",
        employee: {},
        formMessage: "",
      },
    ];
    setValue(data);
    render(<Employee />, container);
  });

  it("Check Employee Save Button", () => {
    const data = [
      {
        form: "save",
        employee: {},
        formMessage: "",
      },
    ];
    setValue(data);
    const wrapper = mount(<Employee />);
    const save = wrapper.find("button").text();
    expect(save).toEqual("Save");
  });

  it("Check Employee Update Button", () => {
    const data = [
      {
        form: "update",
        employee: {},
        formMessage: "",
      },
    ];
    setValue(data);
    const wrapper = mount(<Employee />);
    const update = wrapper.find("button").text();
    expect(update).toEqual("Update");
  });

  it("Check Error Message Displayed", () => {
    const data = [
      {
        form: "save",
        employee: {},
        formMessage: {
          status: 0,
          message: "First Name is a required field",
        },
      },
    ];
    setValue(data);
    const wrapper = mount(<Employee />);
    const h2 = wrapper.find("h2").text();
    expect(h2).toEqual("First Name is a required field");
  });

  it("Check Input Display Employee", () => {
    const data = [
      {
        form: "update",
        employee: {
          employee_id: 1,
          firstName: "Yama",
          lastName: "Cruz",
        },
        formMessage: {},
      },
    ];
    setValue(data);
    const wrapper = mount(<Employee />);
    const firstNameValue = wrapper.find("input").at(0).instance().value;
    const lastNameValue = wrapper.find("input").at(1).instance().value;
    expect(firstNameValue).toEqual("Yama");
    expect(lastNameValue).toEqual("Cruz");
  });

  it("Check Employee Component Test to Match Snapshot Display", () => {
    const data = [
      {
        form: "update",
        employee: {
          employee_id: 1,
          firstName: "Yama",
          lastName: "Cruz",
        },
        formMessage: {},
      },
    ];
    setValue(data);
    const Snapshot = renderer.create(<Employee />).toJSON();
    expect(Snapshot).toMatchSnapshot();
  });
});
