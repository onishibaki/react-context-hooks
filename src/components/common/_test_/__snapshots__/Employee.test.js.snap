// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employee Component Test Check Employee Component Test to Match Snapshot Display 1`] = `
<div
  className="sc-AxjAm EmuRM employee-container"
>
  <div
    className="sc-AxirZ ezqyFp"
  >
    <span
      className="sc-AxhCb cewekA"
      onClick={[Function]}
    >
      X
    </span>
    <label>
      First Name: 
    </label>
    <input
      className="sc-AxhUy gyiYUS"
      defaultValue="Yama"
      name="fname"
      onChange={[Function]}
    />
    <label>
      Last Name: 
    </label>
    <input
      className="sc-AxhUy gyiYUS"
      defaultValue="Cruz"
      name="lname"
      onChange={[Function]}
    />
    <button
      className="sc-AxiKw LINap"
      onClick={[Function]}
    >
      Update
    </button>
  </div>
</div>
`;

exports[`Feedback Component Test Check Employee Component Test to Match Snapshot Display 1`] = `
<div
  className="sc-AxjAm EmuRM employee-container"
>
  <div
    className="sc-AxirZ ezqyFp"
  >
    <span
      className="sc-AxhCb cewekA"
      onClick={[Function]}
    >
      X
    </span>
    <label>
      First Name: 
    </label>
    <input
      className="sc-AxhUy gyiYUS"
      defaultValue="Yama"
      name="fname"
      onChange={[Function]}
    />
    <label>
      Last Name: 
    </label>
    <input
      className="sc-AxhUy gyiYUS"
      defaultValue="Cruz"
      name="lname"
      onChange={[Function]}
    />
    <button
      className="sc-AxiKw LINap"
      onClick={[Function]}
    >
      Update
    </button>
  </div>
</div>
`;
